core/src/main/scala/chisel3/Layer.scala

```diff
@@ -196,6 +197,100 @@ object layer {
     }
   }
 
+  /** A type class that describes how to post-process the return value from a layer block. */
+  sealed trait BlockReturnHandler[A] {
+
+    /** The type that will be returned by this handler. */
+    type R
+
+    /** Do not post-process the return value.  This is what should happen if the
+      * code early exits, i.e., if no layer block is created.
+      */
+    private[chisel3] def identity(result: A): R
+
+    /** Post-process the return value.
+      *
+      * @param placeholder the location above the layer block where commands can be inserted
+      * @param layerBlock the current layer block
+      * @param result the return value of the layer block
+      * @param sourceInfo source locator information
+      * @return the post-processed result
+      */
+    private[chisel3] def apply(placeholder: Placeholder, layerBlock: LayerBlock, result: A)(
+      implicit sourceInfo: SourceInfo
+    ): R
+  }
+
+  object BlockReturnHandler {
+
+    type Aux[A, B] = BlockReturnHandler[A] { type R = B }
+
+    /** Return a [[BlockReturnHandler]] that will always return [[Unit]]. */
+    def unit[A]: Aux[A, Unit] = new BlockReturnHandler[A] {
+
+      override type R = Unit
+
+      override private[chisel3] def identity(result: A): R = ()
+
+      override private[chisel3] def apply(
+        placeholder: Placeholder,
+        layerBlock:  LayerBlock,
+        result:      A
+      )(
+        implicit sourceInfo: SourceInfo
+      ): R = {
+        ()
+      }
+
+    }
+
+    /** Return a [[BlockReturnHandler]] that, if a layer block is created, will
+      * create a [[Wire]] of layer-colored probe type _above_ the layer block
+      * and [[probe.define]] this at the end of the layer block.  If no
+      * layer-block is created, then the result is pass-through.
+      */
+    implicit def layerColoredWire[A <: Data]: Aux[A, A] = new BlockReturnHandler[A] {
+
+      override type R = A
+
+      override private[chisel3] def identity(result: A): R = result
+
+      override private[chisel3] def apply(
+        placeholder: Placeholder,
+        layerBlock:  LayerBlock,
+        result:      A
+      )(
+        implicit sourceInfo: SourceInfo
+      ): R = {
+        // The result wire is either a new layer-colored probe wire or an
+        // existing layer-colored probe wire forwarded up.  If it is the former,
+        // the wire is colored with the current layer.  For the latter, the wire
+        // keeps its existing layer color.
+        val layerColoredWire = placeholder.append {
+          result.probeInfo match {
+            case None            => Wire(probe.Probe(chiselTypeOf(result), layerBlock.layer))
+            case Some(probeInfo) => Wire(chiselTypeOf(result))
+          }
+        }
+
+        // Similarly, if the result is a probe, then we forward it directly.  If
+        // it is a non-probe, then we need to probe if first.
+        Builder.forcedUserModule.withRegion(layerBlock.region) {
```

Comment:
I don't think it matters given the specific hardware created here (?), but just a note that stands out: this creates hardware assuming that changing the block is sufficient (with Placeholder and related builder-state management details fresh on our minds).  I think that's right but is important this only does things where having the rest of the state wrong is not an issue (e.g., layerStack).